File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.js
const fs = require('fs');
const read = fs.readFileSync('dev/stdin').toString().trim();

let [n, arr, oper] = read.split("\n");
arr = arr.split(" ").map(Number);
oper = oper.split(" ").map(Number);

const calcFun = [(a,b) => a + b, (a,b) => a - b, (a,b) => a * b, (a,b) => ~~(a / b)];

const main = () => {
    const result = [-(10 ** 9 + 1), 10 ** 9 + 1];
    const recur = (arr_order, calcv) => {
        if (arr_order === arr.length){
            result[0] = Math.max(calcv, result[0])
            result[1] = Math.min(calcv, result[1])
            return;
        }
        
        for (let i = 0; i < 4; i++){
            if (oper[i] > 0){
                oper[i] -= 1
                recur(arr_order + 1, calcFun[i](calcv, arr[arr_order]));
                oper[i] += 1
            }
        }
    }
    
    recur(1, arr[0])
    
    for(let a of result){
        console.log(a)
    }
}

main()

File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.py
import sys

input = sys.stdin.readline

n = int(input())
arr = [int(i) for i in input().split()]
oper = [int(i) for i in input().split()]
calc_fun = [
    lambda x: x[0] + x[1],
    lambda x: x[0] - x[1],
    lambda x: x[0] * x[1],
    lambda x: int(x[0] / x[1]),
]


def permu():
    li = [float("inf"), -float("inf")]

    def recur(order, calcv):
        if order == len(arr):
            li[0] = min(li[0], calcv)
            li[1] = max(li[1], calcv)
            return

        for i in range(4):
            if oper[i] > 0:
                oper[i] -= 1
                ncalcv = calc_fun[i]((calcv, arr[order]))
                recur(order + 1, ncalcv)
                oper[i] += 1

    recur(1, arr[0])
    return li


def main():
    for a in permu()[::-1]:
        print(a)


main()
